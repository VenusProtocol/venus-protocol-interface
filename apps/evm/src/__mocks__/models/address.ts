const address = '0x3d759121234cd36F8124C21aFe1c6852d2bEd848';

export const altAddress = '0xa258a693A403b7e98fd05EE9e1558C760308cFC7';

export default address;
