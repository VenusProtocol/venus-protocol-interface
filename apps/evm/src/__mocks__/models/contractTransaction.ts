import { ChainId } from 'types';
import type { Transaction } from 'viem';

const contractTransaction: Transaction = {
  hash: '0xfake-transaction-hash',
  type: 'legacy',
  from: '0x1d759121234cd36F8124C21aFe1c6852d2bEd848',
  to: '0x2d759121234cd36F8124C21aFe1c6852d2bEd834',
  nonce: 10,
  value: 0n,
  chainId: ChainId.BSC_TESTNET,
  blockHash: '0xfake-block-hash',
  blockNumber: 1123213n,
  gas: 12n,
  gasPrice: 1n,
  input: '0xfake-input',
  r: '0xfake-r',
  s: '0xfake-s',
  v: 3n,
  transactionIndex: 1,
  typeHex: '0xfake-type-hex',
};

export default contractTransaction;
