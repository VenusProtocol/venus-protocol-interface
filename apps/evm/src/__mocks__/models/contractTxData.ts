import type { ContractTxData } from 'types';

const contractTxData: ContractTxData<any, any> = {
  contract: {},
  methodName: 'fakeMethod',
  args: [],
};

export default contractTxData;
