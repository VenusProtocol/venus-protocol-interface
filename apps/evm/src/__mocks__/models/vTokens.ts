import type { VToken } from 'types';

import { bnb, busd, eth, lisUsd, luna, usdc, usdt, ust, wbnb, weth, xvs } from './tokens';

export const vXvs: VToken = {
  address: '0x6d6F697e34145Bb95c54E77482d97cc261Dc237E',
  decimals: 8,
  symbol: 'vXVS',
  underlyingToken: xvs,
};

export const vBnb: VToken = {
  address: '0x2E7222e51c0f6e98610A1543Aa3836E092CDe62c',
  decimals: 8,
  symbol: 'vBNB',
  underlyingToken: bnb,
};

export const vUsdc: VToken = {
  address: '0xD5C4C2e2facBEB59D0216D0595d63FcDc6F9A1a7',
  decimals: 8,
  symbol: 'vUSDC',
  underlyingToken: usdc,
};

export const vUsdtCorePool: VToken = {
  address: '0xb7526572FFE56AB9D7489838Bf2E18e3323b441A',
  decimals: 8,
  symbol: 'vUSDT',
  underlyingToken: usdt,
};

export const vUsdtDeFi: VToken = {
  address: '0x80CC30811e362aC9aB857C3d7875CbcCc0b65750',
  decimals: 8,
  symbol: 'vUSDT_DeFi',
  underlyingToken: usdt,
};

export const vUsdtGameFi: VToken = {
  address: '0x0bFE4e0B8A2a096A27e5B18b078d25be57C08634',
  decimals: 8,
  symbol: 'vUSDT_GameFi',
  underlyingToken: usdt,
};

export const vUsdtLiquidStakedBnb: VToken = {
  address: '0x2197d02cC9cd1ad51317A0a85A656a0c82383A7c',
  decimals: 8,
  symbol: 'vUSDT_Liquid_Staked_BNB',
  underlyingToken: usdt,
};

export const vUsdtMeme: VToken = {
  address: '0x3AF2bE7AbEF0f840b196D99d79F4B803a5dB14a1',
  decimals: 8,
  symbol: 'vUSDT_Meme',
  underlyingToken: usdt,
};

export const vUsdtStablecoins: VToken = {
  address: '0x3338988d0beb4419Acb8fE624218754053362D06',
  decimals: 8,
  symbol: 'vUSDT_Stablecoins',
  underlyingToken: usdt,
};

export const vUsdtTron: VToken = {
  address: '0x712774CBFFCBD60e9825871CcEFF2F917442b2c3',
  decimals: 8,
  symbol: 'vUSDT_Tron',
  underlyingToken: usdt,
};

export const vBusdCorePool: VToken = {
  address: '0x08e0A5575De71037aE36AbfAfb516595fE68e5e4',
  decimals: 8,
  symbol: 'vBUSD',
  underlyingToken: busd,
};

export const vBusdStablecoins: VToken = {
  address: '0x8c8A1a0b6e1cb8058037F7bF24de6b79Aca5B7B0',
  decimals: 8,
  symbol: 'vBUSD',
  underlyingToken: busd,
};

export const vUst: VToken = {
  address: '0xF206af85BC2761c4F876d27Bd474681CfB335EfA',
  decimals: 8,
  symbol: 'vUST',
  underlyingToken: ust,
};

export const vLuna: VToken = {
  address: '0x9C3015191d39cF1930F92EB7e7BCbd020bCA286a',
  decimals: 8,
  symbol: 'vLUNA',
  underlyingToken: luna,
};

export const vWeth: VToken = {
  address: '0xc2931B1fEa69b6D6dA65a50363A8D75d285e4da9',
  decimals: 8,
  symbol: 'vWETH',
  underlyingToken: weth,
};

export const vEth: VToken = {
  address: '0x86f8DfB7CA84455174EE9C3edd94867b51Da46BD',
  decimals: 8,
  symbol: 'vETH',
  underlyingToken: eth,
};

export const vLisUSD: VToken = {
  address: '0x170d3b2da05cc2124334240fB34ad1359e34C562',
  decimals: 8,
  symbol: 'vlisUSD',
  underlyingToken: lisUsd,
};

export const vWbnb: VToken = {
  address: '0x231dED0Dfc99634e52EE1a1329586bc970d773b3',
  decimals: 8,
  symbol: 'vWBNB',
  underlyingToken: wbnb,
};

export default [
  vXvs,
  vBnb,
  vUsdc,
  vUsdtCorePool,
  vUsdtDeFi,
  vUsdtGameFi,
  vUsdtLiquidStakedBnb,
  vUsdtMeme,
  vUsdtStablecoins,
  vUsdtTron,
  vBusdCorePool,
  vBusdStablecoins,
  vUst,
  vLuna,
  vEth,
  vLisUSD,
  vWbnb,
];
