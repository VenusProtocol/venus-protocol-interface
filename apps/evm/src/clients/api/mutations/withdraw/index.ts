import type BigNumber from 'bignumber.js';
import type { ContractTransaction } from 'ethers';

import type { NativeTokenGateway, VBep20, VBnb } from 'libs/contracts';
import { VError } from 'libs/errors';

export interface WithdrawInput {
  amountMantissa: BigNumber;
  tokenContract?: VBep20 | VBnb;
  nativeTokenGatewayContract?: NativeTokenGateway;
  withdrawFullSupply?: boolean;
  unwrap?: boolean;
}

export type WithdrawOutput = ContractTransaction;

const withdraw = async ({
  amountMantissa,
  tokenContract,
  nativeTokenGatewayContract,
  withdrawFullSupply = false,
  unwrap = false,
}: WithdrawInput): Promise<WithdrawOutput> => {
  // Handle withdraw and unwrap flow
  if (unwrap && !nativeTokenGatewayContract) {
    throw new VError({
      type: 'unexpected',
      code: 'somethingWentWrong',
    });
  }

  if (unwrap) {
    return withdrawFullSupply
      ? nativeTokenGatewayContract!.redeemAndUnwrap(amountMantissa.toFixed())
      : nativeTokenGatewayContract!.redeemUnderlyingAndUnwrap(amountMantissa.toFixed());
  }

  // Handle withdraw flow
  if (!tokenContract) {
    throw new VError({
      type: 'unexpected',
      code: 'somethingWentWrong',
    });
  }

  return withdrawFullSupply
    ? tokenContract.redeem(amountMantissa.toFixed())
    : tokenContract.redeemUnderlying(amountMantissa.toFixed());
};

export default withdraw;
