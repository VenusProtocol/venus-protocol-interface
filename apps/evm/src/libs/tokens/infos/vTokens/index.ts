import type { VTokenAssets } from 'libs/tokens/types';
import { ChainId } from 'types';
import { vTokenAssets as arbitrumOneVTokenAssets } from './arbitrumOne';
import { vTokenAssets as arbitrumSepoliaVTokenAssets } from './arbitrumSepolia';
import { vTokenAssets as baseMainnetVTokenAssets } from './baseMainnet';
import { vTokenAssets as baseSepoliaVTokenAssets } from './baseSepolia';
import { vTokenAssets as berachainBepoliaVTokenAssets } from './berachainBepolia';
import { vTokenAssets as berachainMainnetVTokenAssets } from './berachainMainnet';
import { vTokenAssets as bscMainnetVTokenAssets } from './bscMainnet';
import { vTokenAssets as bscTestnetVTokenAssets } from './bscTestnet';
import { vTokenAssets as ethereumVTokenAssets } from './ethereum';
import { vTokenAssets as opBnbMainnetVTokenAssets } from './opBnbMainnet';
import { vTokenAssets as opBnbTestnetVTokenAssets } from './opBnbTestnet';
import { vTokenAssets as optimismMainnetVTokenAssets } from './optimismMainnet';
import { vTokenAssets as optimismSepoliaVTokenAssets } from './optimismSepolia';
import { vTokenAssets as sepoliaVTokenAssets } from './sepolia';
import { vTokenAssets as unichainMainnetVTokenAssets } from './unichainMainnet';
import { vTokenAssets as unichainSepoliaVTokenAssets } from './unichainSepolia';
import { vTokenAssets as zkSyncMainnetVTokenAssets } from './zkSyncMainnet';
import { vTokenAssets as zkSyncSepoliaVTokenAssets } from './zkSyncSepolia';

export const vTokenAssetsPerChainId: Record<ChainId, VTokenAssets> = {
  [ChainId.BSC_MAINNET]: bscMainnetVTokenAssets,
  [ChainId.BSC_TESTNET]: bscTestnetVTokenAssets,
  [ChainId.OPBNB_MAINNET]: opBnbMainnetVTokenAssets,
  [ChainId.OPBNB_TESTNET]: opBnbTestnetVTokenAssets,
  [ChainId.ETHEREUM]: ethereumVTokenAssets,
  [ChainId.SEPOLIA]: sepoliaVTokenAssets,
  [ChainId.ARBITRUM_ONE]: arbitrumOneVTokenAssets,
  [ChainId.ARBITRUM_SEPOLIA]: arbitrumSepoliaVTokenAssets,
  [ChainId.ZKSYNC_MAINNET]: zkSyncMainnetVTokenAssets,
  [ChainId.ZKSYNC_SEPOLIA]: zkSyncSepoliaVTokenAssets,
  [ChainId.OPTIMISM_MAINNET]: optimismMainnetVTokenAssets,
  [ChainId.OPTIMISM_SEPOLIA]: optimismSepoliaVTokenAssets,
  [ChainId.BASE_MAINNET]: baseMainnetVTokenAssets,
  [ChainId.BASE_SEPOLIA]: baseSepoliaVTokenAssets,
  [ChainId.UNICHAIN_MAINNET]: unichainMainnetVTokenAssets,
  [ChainId.UNICHAIN_SEPOLIA]: unichainSepoliaVTokenAssets,
  [ChainId.BERACHAIN_MAINNET]: berachainMainnetVTokenAssets,
  [ChainId.BERACHAIN_BEPOLIA]: berachainBepoliaVTokenAssets,
};
