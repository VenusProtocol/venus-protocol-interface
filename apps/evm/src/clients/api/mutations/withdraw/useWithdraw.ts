import { queryClient } from 'clients/api';
import withdraw, { type WithdrawInput } from 'clients/api/mutations/withdraw';
import FunctionKey from 'constants/functionKey';
import { type UseSendTransactionOptions, useSendTransaction } from 'hooks/useSendTransaction';
import { useAnalytics } from 'libs/analytics';
import { useGetNativeTokenGatewayContract, useGetVTokenContract } from 'libs/contracts';
import { useChainId } from 'libs/wallet';
import type { VToken } from 'types';
import { callOrThrow, convertMantissaToTokens } from 'utilities';

type TrimmedRedeemInput = Omit<WithdrawInput, 'tokenContract' | 'accountAddress'>;
type Options = UseSendTransactionOptions<TrimmedRedeemInput>;

const useWithdraw = (
  {
    vToken,
    poolName,
    poolComptrollerAddress,
  }: { vToken: VToken; poolName: string; poolComptrollerAddress: string },
  options?: Partial<Options>,
) => {
  const { chainId } = useChainId();
  const nativeTokenGatewayContract = useGetNativeTokenGatewayContract({
    comptrollerContractAddress: poolComptrollerAddress,
    passSigner: true,
  });
  const tokenContract = useGetVTokenContract({
    vToken,
    passSigner: true,
  });

  const { captureAnalyticEvent } = useAnalytics();

  return useSendTransaction({
    fnKey: [FunctionKey.WITHDRAW],
    fn: (input: TrimmedRedeemInput) =>
      callOrThrow({ tokenContract }, params =>
        withdraw({
          ...params,
          ...input,
          nativeTokenGatewayContract,
        }),
      ),
    onConfirmed: async ({ input }) => {
      captureAnalyticEvent('Tokens withdrawn', {
        poolName,
        tokenSymbol: vToken.underlyingToken.symbol,
        tokenAmountTokens: convertMantissaToTokens({
          token: vToken.underlyingToken,
          value: input.amountMantissa,
        }).toNumber(),
        withdrewFullSupply: false,
      });

      const accountAddress = await tokenContract?.signer.getAddress();

      queryClient.invalidateQueries({ queryKey: [FunctionKey.GET_V_TOKEN_BALANCES_ALL] });
      queryClient.invalidateQueries({ queryKey: [FunctionKey.GET_MAIN_MARKETS] });
      queryClient.invalidateQueries({ queryKey: [FunctionKey.GET_LEGACY_POOL] });
      queryClient.invalidateQueries({ queryKey: [FunctionKey.GET_ISOLATED_POOLS] });

      queryClient.invalidateQueries({
        queryKey: [
          FunctionKey.GET_V_TOKEN_BALANCE,
          {
            chainId,
            accountAddress,
            vTokenAddress: vToken.address,
          },
        ],
      });

      queryClient.invalidateQueries({
        queryKey: [
          FunctionKey.GET_TOKEN_BALANCES,
          {
            chainId,
            accountAddress,
          },
        ],
      });

      queryClient.invalidateQueries({
        queryKey: [
          FunctionKey.GET_BALANCE_OF,
          {
            chainId,
            accountAddress,
            tokenAddress: vToken.underlyingToken.address,
          },
        ],
      });

      if (input.unwrap && vToken.underlyingToken.tokenWrapped?.isNative) {
        queryClient.invalidateQueries({
          queryKey: [
            FunctionKey.GET_BALANCE_OF,
            {
              chainId,
              accountAddress,
              tokenAddress: vToken.underlyingToken.tokenWrapped.address,
            },
          ],
        });
      }
    },
    options,
  });
};

export default useWithdraw;
