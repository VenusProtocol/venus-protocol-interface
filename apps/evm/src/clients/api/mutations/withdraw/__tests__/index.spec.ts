import BigNumber from 'bignumber.js';
import type { Address, PublicClient } from 'viem';

import fakeContractTransaction from '__mocks__/models/contractTransaction';
import fakeSigner from '__mocks__/models/signer';
import { vUsdt } from '__mocks__/models/vTokens';
import type { NativeTokenGateway, VBep20, VBnb } from 'libs/contracts';

import withdraw from '..';

const fakeAmount = new BigNumber('10000000000000000');
const fakeAccessList = [{ address: '0xfakeAddress', storageKeys: ['0xfakeStorageKey'] }];
const fakePublicClient = {
  createAccessList: vi.fn(async () => ({
    accessList: fakeAccessList,
    gasUsed: 100000n,
  })),
} as unknown as PublicClient;

describe('withdraw', () => {
  describe('withdraw flow', async () => {
    it('throws an error when vToken contract was not passed', async () => {
      try {
        await withdraw({
          vToken: vUsdt,
          amountMantissa: fakeAmount,
          withdrawFullSupply: true,
          publicClient: fakePublicClient,
        });

        throw new Error('withdraw should have thrown an error but did not');
      } catch (error) {
        expect(error).toMatchInlineSnapshot('[Error: somethingWentWrong]');
      }
    });

    it('returns contract transaction data when request to withdraw full supply succeeds', async () => {
      const redeemMock = vi.fn(async () => fakeContractTransaction);

      const fakeVTokenContract = {
        functions: {
          redeem: redeemMock,
        },
        signer: fakeSigner,
      } as unknown as VBep20;

      const response = await withdraw({
        vToken: vUsdt,
        tokenContract: fakeVTokenContract,
        amountMantissa: fakeAmount,
        withdrawFullSupply: true,
        publicClient: fakePublicClient,
      });

      expect(response).toMatchSnapshot();
    });

    it('returns contract transaction data when request to withdraw partial supply succeeds', async () => {
      const redeemUnderlyingMock = vi.fn(async () => fakeContractTransaction);

      const fakeVTokenContract = {
        functions: {
          redeemUnderlying: redeemUnderlyingMock,
        },
        signer: fakeSigner,
      } as unknown as VBep20;

      const response = await withdraw({
        vToken: vUsdt,
        tokenContract: fakeVTokenContract,
        amountMantissa: fakeAmount,
        publicClient: fakePublicClient,
      });

      expect(response).toMatchSnapshot();
    });

    it('includes accessList when underlying token is native', async () => {
      const redeemUnderlyingMock = vi.fn(async () => fakeContractTransaction);
      const fakeAccountAddress = '0xfakeUserAddress' as Address;
      const getAddressMock = vi.fn(async () => fakeAccountAddress);

      const fakeVTokenContract = {
        functions: {
          redeemUnderlying: redeemUnderlyingMock,
        },
        signer: {
          ...fakeSigner,
          getAddress: getAddressMock,
        },
      } as unknown as VBnb;

      const response = await withdraw({
        vToken: {
          ...vUsdt,
          underlyingToken: {
            ...vUsdt.underlyingToken,
            isNative: true,
          },
        },
        tokenContract: fakeVTokenContract,
        amountMantissa: fakeAmount,
        publicClient: fakePublicClient,
      });

      expect(fakePublicClient.createAccessList).toHaveBeenCalledWith({
        data: '0x',
        value: 1n,
        to: fakeAccountAddress,
      });

      expect(response).toMatchSnapshot();
    });
  });

  describe('withdraw and unwrap flow', async () => {
    it('throws an error when unwrap was passed as true but NativeTokenGateway contract was not passed', async () => {
      try {
        await withdraw({
          vToken: vUsdt,
          amountMantissa: fakeAmount,
          unwrap: true,
          publicClient: fakePublicClient,
        });

        throw new Error('withdraw should have thrown an error but did not');
      } catch (error) {
        expect(error).toMatchInlineSnapshot('[Error: somethingWentWrong]');
      }
    });

    it('returns contract transaction data when request to withdraw full supply succeeds', async () => {
      const redeemAndUnwrapMock = vi.fn(async () => fakeContractTransaction);

      const fakeNativeTokenGatewayContract = {
        functions: {
          redeemAndUnwrap: redeemAndUnwrapMock,
        },
        signer: fakeSigner,
      } as unknown as NativeTokenGateway;

      const response = await withdraw({
        vToken: vUsdt,
        nativeTokenGatewayContract: fakeNativeTokenGatewayContract,
        amountMantissa: fakeAmount,
        withdrawFullSupply: true,
        unwrap: true,
        publicClient: fakePublicClient,
      });

      expect(response).toMatchSnapshot();
    });

    it('returns contract transaction data when request to withdraw partial supply succeeds', async () => {
      const redeemUnderlyingAndUnwrapMock = vi.fn(async () => fakeContractTransaction);

      const fakeNativeTokenGatewayContract = {
        functions: {
          redeemUnderlyingAndUnwrap: redeemUnderlyingAndUnwrapMock,
        },
        signer: fakeSigner,
      } as unknown as NativeTokenGateway;

      const response = await withdraw({
        vToken: vUsdt,
        nativeTokenGatewayContract: fakeNativeTokenGatewayContract,
        amountMantissa: fakeAmount,
        unwrap: true,
        publicClient: fakePublicClient,
      });

      expect(response).toMatchSnapshot();
    });
  });
});
