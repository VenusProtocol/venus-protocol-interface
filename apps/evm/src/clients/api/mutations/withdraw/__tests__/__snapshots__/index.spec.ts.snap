// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`withdraw > withdraw and unwrap flow > returns contract transaction data when request to withdraw full supply succeeds 1`] = `
{
  "args": [
    "10000000000000000",
  ],
  "contract": {
    "functions": {
      "redeemAndUnwrap": [MockFunction spy],
    },
    "signer": {
      "_checkProvider": [MockFunction spy],
      "_isSigner": true,
      "call": [MockFunction spy],
      "checkTransaction": [MockFunction spy],
      "connect": [MockFunction spy],
      "estimateGas": [MockFunction spy],
      "getAddress": [MockFunction spy],
      "getBalance": [MockFunction spy],
      "getChainId": [MockFunction spy],
      "getFeeData": [MockFunction spy],
      "getGasPrice": [MockFunction spy],
      "getTransactionCount": [MockFunction spy],
      "populateTransaction": [MockFunction spy],
      "resolveName": [MockFunction spy],
      "sendTransaction": [MockFunction spy],
      "signMessage": [MockFunction spy],
      "signTransaction": [MockFunction spy],
    },
  },
  "methodName": "redeemAndUnwrap",
}
`;

exports[`withdraw > withdraw and unwrap flow > returns contract transaction data when request to withdraw partial supply succeeds 1`] = `
{
  "args": [
    "10000000000000000",
  ],
  "contract": {
    "functions": {
      "redeemUnderlyingAndUnwrap": [MockFunction spy],
    },
    "signer": {
      "_checkProvider": [MockFunction spy],
      "_isSigner": true,
      "call": [MockFunction spy],
      "checkTransaction": [MockFunction spy],
      "connect": [MockFunction spy],
      "estimateGas": [MockFunction spy],
      "getAddress": [MockFunction spy],
      "getBalance": [MockFunction spy],
      "getChainId": [MockFunction spy],
      "getFeeData": [MockFunction spy],
      "getGasPrice": [MockFunction spy],
      "getTransactionCount": [MockFunction spy],
      "populateTransaction": [MockFunction spy],
      "resolveName": [MockFunction spy],
      "sendTransaction": [MockFunction spy],
      "signMessage": [MockFunction spy],
      "signTransaction": [MockFunction spy],
    },
  },
  "methodName": "redeemUnderlyingAndUnwrap",
}
`;

exports[`withdraw > withdraw flow > includes accessList when underlying token is native 1`] = `
{
  "args": [
    "10000000000000000",
  ],
  "contract": {
    "functions": {
      "redeemUnderlying": [MockFunction spy],
    },
    "signer": {
      "_checkProvider": [MockFunction spy],
      "_isSigner": true,
      "call": [MockFunction spy],
      "checkTransaction": [MockFunction spy],
      "connect": [MockFunction spy],
      "estimateGas": [MockFunction spy],
      "getAddress": [MockFunction spy] {
        "calls": [
          [],
        ],
        "results": [
          {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "getBalance": [MockFunction spy],
      "getChainId": [MockFunction spy],
      "getFeeData": [MockFunction spy],
      "getGasPrice": [MockFunction spy],
      "getTransactionCount": [MockFunction spy],
      "populateTransaction": [MockFunction spy],
      "resolveName": [MockFunction spy],
      "sendTransaction": [MockFunction spy],
      "signMessage": [MockFunction spy],
      "signTransaction": [MockFunction spy],
    },
  },
  "methodName": "redeemUnderlying",
  "overrides": {
    "accessList": [
      {
        "address": "0xfakeAddress",
        "storageKeys": [
          "0xfakeStorageKey",
        ],
      },
    ],
  },
}
`;

exports[`withdraw > withdraw flow > returns contract transaction data when request to withdraw full supply succeeds 1`] = `
{
  "args": [
    "10000000000000000",
  ],
  "contract": {
    "functions": {
      "redeem": [MockFunction spy],
    },
    "signer": {
      "_checkProvider": [MockFunction spy],
      "_isSigner": true,
      "call": [MockFunction spy],
      "checkTransaction": [MockFunction spy],
      "connect": [MockFunction spy],
      "estimateGas": [MockFunction spy],
      "getAddress": [MockFunction spy],
      "getBalance": [MockFunction spy],
      "getChainId": [MockFunction spy],
      "getFeeData": [MockFunction spy],
      "getGasPrice": [MockFunction spy],
      "getTransactionCount": [MockFunction spy],
      "populateTransaction": [MockFunction spy],
      "resolveName": [MockFunction spy],
      "sendTransaction": [MockFunction spy],
      "signMessage": [MockFunction spy],
      "signTransaction": [MockFunction spy],
    },
  },
  "methodName": "redeem",
  "overrides": undefined,
}
`;

exports[`withdraw > withdraw flow > returns contract transaction data when request to withdraw partial supply succeeds 1`] = `
{
  "args": [
    "10000000000000000",
  ],
  "contract": {
    "functions": {
      "redeemUnderlying": [MockFunction spy],
    },
    "signer": {
      "_checkProvider": [MockFunction spy],
      "_isSigner": true,
      "call": [MockFunction spy],
      "checkTransaction": [MockFunction spy],
      "connect": [MockFunction spy],
      "estimateGas": [MockFunction spy],
      "getAddress": [MockFunction spy],
      "getBalance": [MockFunction spy],
      "getChainId": [MockFunction spy],
      "getFeeData": [MockFunction spy],
      "getGasPrice": [MockFunction spy],
      "getTransactionCount": [MockFunction spy],
      "populateTransaction": [MockFunction spy],
      "resolveName": [MockFunction spy],
      "sendTransaction": [MockFunction spy],
      "signMessage": [MockFunction spy],
      "signTransaction": [MockFunction spy],
    },
  },
  "methodName": "redeemUnderlying",
  "overrides": undefined,
}
`;
